ibtk/src/utilities/HierarchyIntegrator.cpp

```diff
@@ -195,6 +195,18 @@ HierarchyIntegrator::initializePatchHierarchy(Pointer<PatchHierarchy<NDIM> > hie
         const int coarsest_ln = 0;
         const int finest_ln = d_hierarchy->getFinestLevelNumber();
         d_gridding_alg->getTagAndInitializeStrategy()->resetHierarchyConfiguration(d_hierarchy, coarsest_ln, finest_ln);
+
+        // Reset hierarchy dependent data for all integrators.
+        hier_integrators.push_back(this);
+        while (!hier_integrators.empty())
+        {
+            HierarchyIntegrator* integrator = hier_integrators.front();
+            integrator->regridHierarchyEndSpecialized();
+            // Add child integrators to the end of the list
+            hier_integrators.pop_front();
+            hier_integrators.insert(
+                hier_integrators.end(), integrator->d_child_integrators.begin(), integrator->d_child_integrators.end());
+        }
```

Comment:
AFAICT this results in calling `regridHierarchyEndSpecialized()` without ever calling `regridHierarchyBeginSpecialized()` - we assume that these are always called in pairs so this isn't going to work for some things (like IBFE). In fact, this does weird things with how we do the divergence-free projection for INS, which causes lots of test failures.

Reply:
A good example of how this is problematic is one of the IBFE tests:
```
        IBFE/explicit_ex4_2d.multilevel.a.input --------------------- RUN FAILED
test failed with output:
    P=00000:Program abort called in file ``/__w/IBAMR/IBAMR/ibtk/src/lagrangian/FEDataManager.cpp'' at line 2935
    P=00000:ERROR MESSAGE: 
    P=00000:Failed assertion: d_eulerian_data_cache->getCoarsestLevelNumber() == 0
```

Reply:
This is not paired because there is no hierarchy at that point. I think it will cause error because we will be deallocating operators that do not exist. 

We do need to fix this issue because the current way of doing things do not consider the scenario where number of levels can change in the simulation. An AMR library should allow for variable number of levels.  I do not know what is the best way of doing it. What I did is makes the variable number of levels work (I tested on a phase change problem), but it has some side effects that I don't know how to fix. Nevertheless, it gives a possible means to address this issue. 

Would it be possible to do some further digging and debugging for IBFE tests? Maybe that will suggest further changes.

Reply:
I did some investigating. It looks like IBFEMethod expects certain objects to be set up by the time that function is called. In particular, with the secondary hierarchy turned on, it expects some values to be cached by `begin` and then read again by `end` - calling `end` alone is not going to work.

The answer is to clearly delineate tasks between `resetHierarchyConfigurationSpecialized()` and the `regridHierarchyBeginSpecialized()` / `regridHierarchyEndSpecialized()` pair. The pair must be called together, otherwise we are going to break things. One further source of problems here is staged initialization: objects go through a sequence of steps in which they are not ready for use until the end, which makes it harder to figure out who is responsible for what. We should avoid booleans like `is_initialized` or `is_allocated` as much as possible and find cleaner ways to enforce invariants (we want one possible state, not four).

Hence - the next action is to compile notes on the responsibilities of these three functions. Once that is done we can sort out what should go where.